nomad/structs/services.go

```diff
@@ -269,17 +277,110 @@ func (sc *ServiceCheck) validate() error {
 		return fmt.Errorf("invalid address_mode %q", sc.AddressMode)
 	}
 
-	// Validate OnUpdate
+	// validate on_update
 	switch sc.OnUpdate {
 	case "", OnUpdateIgnore, OnUpdateRequireHealthy, OnUpdateIgnoreWarn:
 		// OK
 	default:
 		return fmt.Errorf("on_update must be %q, %q, or %q; got %q", OnUpdateRequireHealthy, OnUpdateIgnoreWarn, OnUpdateIgnore, sc.OnUpdate)
 	}
 
+	// validate check_restart and on_update do not conflict
+	if sc.CheckRestart != nil {
+		// CheckRestart and OnUpdate Ignore are incompatible If OnUpdate treats
+		// an error has healthy, and the deployment succeeds followed by check
+		// restart restarting failing checks, the deployment is left in an odd
+		// state
+		if sc.OnUpdate == OnUpdateIgnore {
+			return fmt.Errorf("on_update value %q is not compatible with check_restart", sc.OnUpdate)
+		}
+		// CheckRestart IgnoreWarnings must be true if a check has defined OnUpdate
+		// ignore_warnings
+		if !sc.CheckRestart.IgnoreWarnings && sc.OnUpdate == OnUpdateIgnoreWarn {
+			return fmt.Errorf("on_update value %q not supported with check_restart ignore_warnings value %q", sc.OnUpdate, strconv.FormatBool(sc.CheckRestart.IgnoreWarnings))
+		}
+	}
+
+	// validate check_restart
+	if err := sc.CheckRestart.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validate a Service's ServiceCheck in the context of the Nomad provider.
+func (sc *ServiceCheck) validateNomad() error {
+	allowable := []string{ServiceCheckTCP, ServiceCheckHTTP}
+	if err := sc.validateCommon(allowable); err != nil {
+		return err
+	}
+
+	// expose is connect (consul) specific
+	if sc.Expose {
+		return fmt.Errorf("expose may only be set for Consul service checks")
+	}
+
+	// nomad checks do not have warnings
+	if sc.OnUpdate == "ignore_warnings" {
```

Comment:
I wish we used more string consts... That `s` on the end would be easy to forget. Nothing we need to block this PR for though.

Reply:
I'll do this in a followup PR